src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -1805,13 +1805,23 @@ AccessExpressionSyntax access when access.BaseExpression is ForSyntax
                         throw new InvalidOperationException("Expected ImportWithClauseSyntax to have a parent.");
                     }
 
-                    if (GetDeclaredTypeAssignment(parent) is not { } extensionAssignment ||
-                        extensionAssignment.Reference.Type is not NamespaceType namespaceType)
+                    ObjectType? configType = null;
+
+                    if (GetDeclaredTypeAssignment(parent) is not { } extensionAssignment)
                     {
                         return null;
                     }
 
-                    if (namespaceType.ConfigurationType is null)
+                    if (extensionAssignment.Reference.Type is NamespaceType namespaceType)
+                    {
+                        configType = namespaceType.ConfigurationType;
+                    }
+                    else if (parent is ExtensionConfigAssignmentSyntax && extensionAssignment.Reference.Type is ObjectType configTypeFromAssignment)
+                    {
+                        configType = configTypeFromAssignment;
```

Comment:
Could you add a comment to explain what this branch is doing? This is pretty hard to follow, and generally having the declared type depend on an assigned type is something we want to avoid, as it leads to cycles. (I assume the type of the extension syntax in a bicepparam file is somehow derived from the assigned type of the config in the associated .bicep file?)

Reply:
I think this depends on the declared type? Honestly having a bit of trouble delineating myself.
The namespace type is the declared type provided by the bicep file, and this syntax is only available in bicepparam files. The config object type is a child of the declared namespace type.

Reply:
Talked about this offline. I think there was confusion about the naming of certain types. `DeclaredTypeAssignment` is slightly confusing in naming (alternative names could be `SyntaxDeclaredTypeMapping` or `DeclaredTypeMapping`). There's also `assignment` in the type `ExtensionConfigAssignmentSyntax` which led to the variable name `configTypeFromAssignment`. Things should be OK here because the type is sourced from another semantic model.